### ISSUE TITLE
double free or corruption when indexing a variable (oddly corrupted file?)

### ISSUE THREAD

**Issue by 5tefan** [Unidata/netcdf4-python#991]
Hello,

I believe I need help diagnosing this potentially corrupted file. The reason I suspect the file may be corrupted is that `file in question -> ncdump -> ncgen -> new file` produces a valid file that does not display this issue. The complicating factor is that the issue with the original file occurs with netcdf-python version 1.5.2 and 1.5.3, but NOT in version 1.4.2 (I have not tested other intermediary versions).  

The following code and attached file cause a problem (output shown below when using Python 3.8.1 with netcdf-python 1.5.3 and netcdf4libversion 4.6.3). 

Code:
```
import netCDF4 as nc
import sys

with nc.Dataset(sys.argv[1]) as nc_in:
    var = "method"
    print(f"Examining variable {var}...")
    print(f"{nc_in.variables[var]}")
    print(f"Item 0: {nc_in.variables[var][0]}")
    print(f"Item 1: {nc_in.variables[var][1]}")
```

File: [dn_magn-l2-hires_g16_s20170414T192300Z_e20170414T192359Z_v2_0_0.zip](https://github.com/Unidata/netcdf4-python/files/4094564/dn_magn-l2-hires_g16_s20170414T192300Z_e20170414T192359Z_v2_0_0.zip)

The file appears valid by inspection with ncdump netcdf library version 4.3.3.1 of Dec 10 2015 16:44:18. I also checked it with h5dump. Yet the simple code above produces the following output:

```
Examining variable method...
<class 'netCDF4._netCDF4.Variable'>
uint8 method(time)
    long_name: magnetic_field_estimation_method
    comments: Method used to achieve best ambient estimate (variables named b_*). Can be {inboard, outboard, average, gradiometer}.
    _FillValue: 255
    valid_min: 0
    valid_max: 3
    flag_values: [0 1 2 3]
    flag_masks: [255 255 255 255]
    flag_names: inboard outboard average gradiometer
unlimited dimensions: time
current shape = (600,)
filling on
Item 0: 0.0
*** Error in `python': double free or corruption (out): 0x000055636f222e30 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x81679)[0x7fb78ae94679]
/home/NESDIS/stefan.codrescu/anaconda3/envs/latest/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-x86_64-linux-gnu.so(+0x2a87c)[0x7fb781ef187c]
/home/NESDIS/stefan.codrescu/anaconda3/envs/latest/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-x86_64-linux-gnu.so(+0x2dfe3)[0x7fb781ef4fe3]
/home/NESDIS/stefan.codrescu/anaconda3/envs/latest/lib/python3.8/site-packages/netCDF4/_netCDF4.cpython-38-x86_64-linux-gnu.so(+0x8f970)[0x7fb783cf7970]
python(PyObject_GetItem+0x49)[0x55636d7d5d89]
python(_PyEval_EvalFrameDefault+0xb4a)[0x55636d8432fa]
python(_PyEval_EvalCodeWithName+0x2e4)[0x55636d7ef5f4]
python(PyEval_EvalCodeEx+0x44)[0x55636d7f02d4]
python(PyEval_EvalCode+0x1c)[0x55636d7f02fc]
python(+0x231944)[0x55636d890944]
python(+0x25de74)[0x55636d8bce74]
python(PyRun_FileExFlags+0xa1)[0x55636d8bf6f1]
python(PyRun_SimpleFileExFlags+0x1c1)[0x55636d8bf8e1]
python(Py_RunMain+0x36d)[0x55636d8c04ed]
python(Py_BytesMain+0x39)[0x55636d8c06e9]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x7fb78ae35505]
python(+0x201c13)[0x55636d860c13]
======= Memory map: ========
55636d65f000-55636d6bd000 r--p 000000
[...]
Aborted (core dumped)
```

**Comment by jswhit** [comment]
Can't reproduce here with netcdf4-python master and python 3.7.4

```
netcdf4-python version: 1.5.4
HDF5 lib version:       1.10.5
netcdf lib version:     4.7.0
numpy version           1.17.1
cython version           0.29.13
```

```python
>>> from netCDF4 import Dataset
>>> nc = Dataset('dn_magn-l2-hires_g16_s20170414T192300Z_e20170414T192359Z_v2_0_0.nc')
>>> print(nc['method'])

<class 'netCDF4._netCDF4.Variable'>
uint8 method(time)
    long_name: magnetic_field_estimation_method
    comments: Method used to achieve best ambient estimate (variables named b_*). Can be {inboard, outboard, average, gradiometer}.
    _FillValue: 255
    valid_min: 0
    valid_max: 3
    flag_values: [0 1 2 3]
    flag_masks: [255 255 255 255]
    flag_names: inboard outboard average gradiometer
unlimited dimensions: time
current shape = (600,)
filling on
```

**Comment by 5tefan** [comment]
Interesting, thanks. I will try to recreate your configuration and report back.

**Comment by 5tefan** [comment]
Very interesting. I agree with you, the issue does not reproduce on master at 7a08c4e, with Python 3.7.4 or 3.8.1. I tried again with NetCDF-Python 1.5.3 (latest) from PyPI, which still reproduces the issue for me on Python 3.7.4. I had a brief look through the changelog and commit history but didn't see what would obviously impact this. I also noticed that the netcdf4libversion is the same 4.4.1.1 under NetCDF-Python 1.5.3 and master... so it shouldn't be something in the underlying C lib? I suppose this makes me hopeful that 1.5.4 will be released soon.
